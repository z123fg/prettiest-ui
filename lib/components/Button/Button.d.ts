import React, { FC } from "react";
type ButtonVariants = "contained" | "outlined" | "text";
export interface ButtonProps extends React.ComponentPropsWithoutRef<"button"> {
    /**
     * Used to decide the variant of the appearance
     **/
    variant?: ButtonVariants;
    /**
     * Used to decide the color of the appearance
     **/
    color?: "primary" | "secondary" | "default";
    /**
     * Used to decide the size of the button
     **/
    size?: "small" | "medium" | "large";
    /**
     * Used to indicate the button is loading
     **/
    isLoading?: boolean;
}
declare const Button: FC<ButtonProps>;
export default Button;
