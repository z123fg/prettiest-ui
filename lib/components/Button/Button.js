var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
};
var __spreadArray = (this && this.__spreadArray) || function (to, from, pack) {
    if (pack || arguments.length === 2) for (var i = 0, l = from.length, ar; i < l; i++) {
        if (ar || !(i in from)) {
            if (!ar) ar = Array.prototype.slice.call(from, 0, i);
            ar[i] = from[i];
        }
    }
    return to.concat(ar || Array.prototype.slice.call(from));
};
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { useState, useEffect } from "react";
var counter = 0;
var Button = function (_a) {
    var _b = _a.color, color = _b === void 0 ? "primary" : _b, _c = _a.variant, variant = _c === void 0 ? "contained" : _c, children = _a.children, _d = _a.size, size = _d === void 0 ? "medium" : _d, _e = _a.disabled, disabled = _e === void 0 ? false : _e, onClick = _a.onClick, className = _a.className, rest = __rest(_a, ["color", "variant", "children", "size", "disabled", "onClick", "className"]);
    var _f = useState(null), cursorPos = _f[0], setCursorPos = _f[1];
    var _g = useState([]), ripples = _g[0], setRipples = _g[1];
    useEffect(function () {
        //create a ripple
        if (cursorPos === null)
            return;
        var newRipple = (_jsx("div", { className: "btn-ripple-".concat(color, "-").concat(variant), style: {
                position: "absolute",
                top: cursorPos.y,
                left: cursorPos.x,
                transform: "translate(-50%, -50%)",
            }, onAnimationEnd: function () {
                setRipples(function (prev) {
                    var nextRipples = __spreadArray([], prev, true);
                    nextRipples.shift();
                    return nextRipples;
                });
            } }, counter++));
        setRipples(function (prev) { return __spreadArray(__spreadArray([], prev, true), [newRipple], false); });
    }, [cursorPos]);
    var handleClick = function (e) {
        var _a = e.nativeEvent, offsetX = _a.offsetX, offsetY = _a.offsetY;
        setCursorPos({ x: offsetX, y: offsetY });
        onClick === null || onClick === void 0 ? void 0 : onClick(e);
    };
    var constructClassName = function () {
        var colorVariantCls = "btn-".concat(color, "-").concat(variant);
        var sizeCls = "btn-".concat(size);
        return ["btn", colorVariantCls, sizeCls].join(" ");
    };
    return (_jsxs("button", __assign({ disabled: disabled, className: constructClassName() + (className ? " " + className : ""), onClick: handleClick }, rest, { children: [children, _jsx("span", { children: ripples })] })));
};
export default Button;
//# sourceMappingURL=Button.js.map