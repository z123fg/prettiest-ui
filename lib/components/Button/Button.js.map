{"version":3,"file":"Button.js","sourceRoot":"","sources":["../../../src/components/Button/Button.tsx"],"names":[],"mappings":";;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;AAAA,OAAc,EAAgD,QAAQ,EAAE,SAAS,EAAE,MAAM,OAAO,CAAC;AAyCjG,IAAI,OAAO,GAAG,CAAC,CAAC;AAChB,IAAM,MAAM,GAAoB,UAAC,EAShC;IARG,IAAA,aAAiB,EAAjB,KAAK,mBAAG,SAAS,KAAA,EACjB,eAAqB,EAArB,OAAO,mBAAG,WAAW,KAAA,EACrB,QAAQ,cAAA,EACR,YAAe,EAAf,IAAI,mBAAG,QAAQ,KAAA,EACf,gBAAgB,EAAhB,QAAQ,mBAAG,KAAK,KAAA,EAChB,OAAO,aAAA,EACP,SAAS,eAAA,EACN,IAAI,cARsB,4EAShC,CADU;IAED,IAAA,KAA4B,QAAQ,CAAmB,IAAI,CAAC,EAA3D,SAAS,QAAA,EAAE,YAAY,QAAoC,CAAC;IAC7D,IAAA,KAAwB,QAAQ,CAAc,EAAE,CAAC,EAAhD,OAAO,QAAA,EAAE,UAAU,QAA6B,CAAC;IACxD,SAAS,CAAC;QACN,iBAAiB;QACjB,IAAI,SAAS,KAAK,IAAI;YAAE,OAAO;QAC/B,IAAM,SAAS,GAAG,CACd,cAEI,SAAS,EAAE,qBAAc,KAAK,cAAI,OAAO,CAAE,EAC3C,KAAK,EAAE;gBACH,QAAQ,EAAE,UAAU;gBACpB,GAAG,EAAE,SAAS,CAAC,CAAC;gBAChB,IAAI,EAAE,SAAS,CAAC,CAAC;gBACjB,SAAS,EAAE,uBAAuB;aACrC,EACD,cAAc,EAAE;gBACZ,UAAU,CAAC,UAAC,IAAI;oBACZ,IAAI,WAAW,qBAAO,IAAI,OAAC,CAAC;oBAC5B,WAAW,CAAC,KAAK,EAAE,CAAC;oBACpB,OAAO,WAAW,CAAC;gBACvB,CAAC,CAAC,CAAC;YACP,CAAC,IAdI,OAAO,EAAE,CAeX,CACV,CAAC;QACF,UAAU,CAAC,UAAC,IAAI,IAAK,uCAAI,IAAI,UAAE,SAAS,WAAnB,CAAoB,CAAC,CAAC;IAC/C,CAAC,EAAE,CAAC,SAAS,CAAC,CAAC,CAAC;IAChB,IAAM,WAAW,GAAG,UAAC,CAAgC;QAC3C,IAAA,KAAuB,CAAC,CAAC,WAAW,EAAlC,OAAO,aAAA,EAAE,OAAO,aAAkB,CAAC;QAC3C,YAAY,CAAC,EAAE,CAAC,EAAE,OAAO,EAAE,CAAC,EAAE,OAAO,EAAE,CAAC,CAAC;QACzC,OAAO,aAAP,OAAO,uBAAP,OAAO,CAAG,CAAC,CAAC,CAAC;IACjB,CAAC,CAAC;IAEF,IAAM,kBAAkB,GAAiB;QACrC,IAAM,eAAe,GAAG,cAAO,KAAK,cAAI,OAAO,CAAE,CAAC;QAClD,IAAM,OAAO,GAAG,cAAO,IAAI,CAAE,CAAC;QAC9B,OAAO,CAAC,KAAK,EAAE,eAAe,EAAE,OAAO,CAAC,CAAC,IAAI,CAAC,GAAG,CAAC,CAAC;IACvD,CAAC,CAAC;IAEF,OAAO,CACH,2BACI,QAAQ,EAAE,QAAQ,EAClB,SAAS,EAAE,kBAAkB,EAAE,GAAG,CAAC,SAAS,CAAC,CAAC,CAAC,GAAG,GAAG,SAAS,CAAC,CAAC,CAAC,EAAE,CAAC,EACpE,OAAO,EAAE,WAAW,IAChB,IAAI,eAEP,QAAQ,EACT,yBAAO,OAAO,GAAQ,KACjB,CACZ,CAAC;AACN,CAAC,CAAC;AAEF,eAAe,MAAM,CAAC"}