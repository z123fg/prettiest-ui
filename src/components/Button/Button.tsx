import React, { FC, FunctionComponent, MouseEvent, ReactNode } from 'react'

type ButtonVariants = "contained" | "outlined" | "text";


interface ButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  variant?: ButtonVariants;
  color?:"primary"|"secondary" | "default";
  size?:"small" | "medium" | "large";
  isLoading?:boolean;
}

//type extendedButtonProps = ButtonProps & React.ButtonHTMLAttributes<HTMLButtonElement> 

const Button:FC<ButtonProps> = ({color="primary", variant="contained",children,size="medium", disabled=false, onClick, className,...rest}) => {

  const handleClick = (e: MouseEvent<HTMLButtonElement>) => {
    onClick?.(e);
  }

  const constructClassName:()=>string = () => {
    const colorVariantCls = `btn-${color}-${variant}`;
    const sizeCls = `btn-${size}`
    return ["btn", colorVariantCls, sizeCls].join(" ")
  }

  return (
    <button disabled={disabled} className={constructClassName()+ " " + className}  onClick={handleClick} {...rest}>
      {children}
    </button>
  )
}

export default Button