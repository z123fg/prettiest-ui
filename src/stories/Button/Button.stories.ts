import type { Meta, StoryObj } from '@storybook/react';
import Button from "../../components/Button/Button"

const meta = {
  title: 'Example/Button',
  component: Button,
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
  tags: ['autodocs'],
} satisfies Meta<typeof Button>;

export default meta;
type Story = StoryObj<typeof meta>;

export const Outlined: Story = {
  args: {
    variant:"outlined",
    children:"Submit"
  },
};
export const Text: Story = {
    args: {
      variant:"text",
      children:"Submit"
    },
  };

  export const Large: Story = {
    args: {
      size:"large",
      children:"Submit"
    },
  };

  export const Small: Story = {
    args: {
      size:"small",
      children:"Submit"
    },
  };

  export const Secondary: Story = {
    args: {
      color:"secondary",
      children:"Submit"
    },
  };
  export const Default: Story = {
    args: {
      color:"default",
      children:"Submit"
    },
  };