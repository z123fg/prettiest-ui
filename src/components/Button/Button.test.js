import { render, screen, fireEvent } from "@testing-library/react";
import Button from "./Button";

describe("button component", () => {
    test("snapshot test", () => {
        const { asFragment } = render(<Button>Submit!</Button>);
        expect(asFragment()).toMatchSnapshot();
    })
    test("button component should be customized by passing variant props", () => {
        const { rerender } = render(<Button>Submit</Button>);
        const buttonEl = screen.queryByText("Submit");
        expect(buttonEl).toHaveClass("btn-primary-contained");
        rerender(<Button variant="outlined">Submit</Button>);
        expect(buttonEl).toHaveClass("btn-primary-outlined");
    });
    test("button component should be customized by passing color props", () => {
        const { rerender } = render(<Button>Submit</Button>);
        const buttonEl = screen.queryByText("Submit");
        expect(buttonEl).toHaveClass("btn-primary-contained");
        rerender(<Button color="secondary">Submit</Button>);
        expect(buttonEl).toHaveClass("btn-secondary-contained");
    });
    test("button component should be customized by passing size props", () => {
        const { rerender } = render(<Button>Submit</Button>);
        const buttonEl = screen.queryByText("Submit");
        expect(buttonEl).toHaveClass("btn-medium");
        rerender(<Button size="large">Submit</Button>);
        expect(buttonEl).toHaveClass("btn-large");
    });

    test("button should invoke the onClick props when it is clicked", () => {
        const mockFn = jest.fn();
        const { rerender } = render(<Button onClick={mockFn}>Submit</Button>);
        expect(mockFn).toBeCalledTimes(0);
        const buttonEl =  screen.queryByText("Submit");
        fireEvent.click(buttonEl);
        expect(mockFn).toBeCalledTimes(1);
    });
});
