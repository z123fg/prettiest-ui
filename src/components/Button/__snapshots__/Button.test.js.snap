// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`button component snapshot test 1`] = `
<DocumentFragment>
  <button
    class="btn btn-primary-contained btn-medium"
  >
    Submit!
  </button>
</DocumentFragment>
`;
